### ISSUE TITLE
`ArgumentError: +bind+ is not a valid predicate name` - possible deeper issue with Check#method_missing?

### ISSUE THREAD

**Issue by georgemillo** [dry-rb/dry-validation#273]
Here's my schema:

```
schema = Dry::Validation.Schema do
  optional(:closed_at).maybe
  optional(:closed).filled(:bool?)
  required(:opened_at).filled(:date?)

  rule(closed_at_must_be_filled: [:closed, :closed_at]) do |closed, closed_at|
    closed.true? > closed_at.filled? & closed_at.date?
  end

  rule(opened_before_closed: [:closed, :closed_at, :opened_at]) do |closed, closed_at, opened_at|
    closed.true? > closed_at.filled? & date? & gteq?(opened_at)
  end
end
```

In plain English: opened_at and closed_at are both Dates. Opened_at must always be filled. If `closed?` is false, closed_at must be empty; else closed_at must be filled AND must not be earlier than opened_at.

Except I must be doing something wrong, because the above code crashed with this error message (note that I get this error when trying to create the schema, not when calling `call`).

```
ArgumentError: +bind+ is not a valid predicate name
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/predicate_registry.rb:109:in `raise_unknown_predicate_error'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/predicate_registry.rb:102:in `ensure_valid_predicate'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema/check.rb:26:in `method_missing'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule.rb:76:in `block in eval_args'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule.rb:76:in `map'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule.rb:76:in `eval_args'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema_compiler.rb:50:in `visit_predicate'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:21:in `visit'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema_compiler.rb:46:in `visit_rule'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:21:in `visit'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:65:in `visit_and'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:21:in `visit'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema_compiler.rb:46:in `visit_rule'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:21:in `visit'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:80:in `visit_implication'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:21:in `visit'
... 3 levels...
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:21:in `visit'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema_compiler.rb:46:in `visit_rule'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-logic-0.4.0/lib/dry/logic/rule_compiler.rb:21:in `visit'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema.rb:117:in `block in initialize_checks'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema.rb:116:in `each'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema.rb:116:in `each_with_object'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema.rb:116:in `initialize_checks'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema.rb:55:in `initialize'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema/class_interface.rb:34:in `new'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation/schema/class_interface.rb:34:in `new'
	from /Users/george/.rvm/gems/ruby-2.3.0/gems/dry-validation-0.10.3/lib/dry/validation.rb:28:in `Schema'
```

I'm not sure whether this is a bug or if I've just fucked something up in how I've created the schema, but on closer inspection, [this line in dry-logic](https://github.com/dry-rb/dry-logic/blob/v0.4.0/lib/dry/logic/rule.rb#L76) is where the problems start:

```
      def eval_args(object)
        with(args: args.map { |arg| arg.instance_of?(UnboundMethod) ? arg.bind(object).() : arg })
      end
```

At one point, `arg` will be an instance of `Dry::Validation::Schema::Check`. Since `Check` is a `BasicObject`, it doesn't respond to `instance_of?`, so `arg.instance_of?(UnboundMethod)` will trigger `method_missing`. `Check#method_missing` then returns a truthy value, so `arg.bind(object)` gets called, triggering `method_missing` again with `:bind` as the method name. `:bind` then gets passed around a few times before eventually triggering the `not a valid predicate name` error.

So my questions are:

1. Is this is a bug, or is there something wrong with my schema design?
2. If it's not a bug, can things be altered so that when n00bs like me try to define an invalid schema we see a more helpful error message?
3. `eval_args` clearly doesn't expect to be passed an instance of `Check`. Could the fact that it is being passed one be indicative of some deeper issue?

Thanks!

**Comment by solnic** [comment]
Here's how it should be written:

``` ruby
schema = Dry::Validation.Schema do
  optional(:closed_at).maybe(:date?)
  optional(:closed).filled(:bool?)
  required(:opened_at).filled(:date?)

  rule(closed_at_must_be_filled: [:closed, :closed_at]) do |closed, closed_at|
    closed.true? > closed_at.date?
  end

  rule(opened_before_closed: [:closed, :closed_at, :opened_at]) do |closed, closed_at, opened_at|
    closed.true? > closed_at.gteq?(opened_at)
  end
end
```

This inconsistency in `rule` DSL is unfortunate, this will be changed before 1.0.0

**Comment by georgemillo** [comment]
So what's the specific error I made here? Calling `.filled?` inside a `rule` block?

It seems strange to me that the 'correct' way of doing this involves calling `date?` twice on the same key (`closed_at`, on line 2 of your code and on line 7). Is this what will be changed in 1.0.0?

**Comment by solnic** [comment]
What will change is that the syntax will be the same as in the main dsl, so ie `true?(closed) > date?(closed_at)`.

The error was caused by calling `date?` w/o arguments which caused the dsl to try to resolve it as a dynamic method that is later called on a schema object, of course the error is misleading in that case, which we can improve.

Re calling `date?` twice - yes, this could be written as `maybe(:date?)` and then you can just check `closed_at.filled?` in the high-level rule, like you originally did.

**Comment by georgemillo** [comment]
Got it. Thanks!
